middle_end/flambda2/types/env/typing_env.ml

```diff
@@ -714,15 +714,42 @@ and add_equation1 t name ty ~(meet_type : meet_type) =
               "Directly recursive equation@ %a = %a@ disallowed:@ %a" Name.print
               name TG.print ty print t)
         ~const:(fun _ -> ()));
-  let simple, t, ty =
-    let aliases = aliases t in
+  let find_canonical name =
+    let aliases =
+      (* If we already have an equation on [name], then we already have the
+         correct alias relation in the current [aliases]. *)
+      if Name.Map.mem name (names_to_types t)
+      then aliases t
+      else
+        (* If we don't have any equation yet, then we can find the canonical
+           element by looking in the corresponding compilation unit *)
+        let comp_unit = Name.compilation_unit name in
+        if Compilation_unit.equal comp_unit
+             (Compilation_unit.get_current_exn ())
+        then aliases t
+        else
+          match (resolver t) comp_unit with
+          | None ->
+            (* The corresponding cmx is unavailable, so even if [name] was not
+               canonical in its compilation unit we won't ever discover it.
+               Relying on the current aliases structure is fine. *)
```

Comment:
Could this go wrong if another compilation unit down the line _does_ have access to the cmx?

Reply:
That is a good question (please augment the comment with the answer)...

Reply:
Yes, there seems to be a potential problem. It could be fixed by recursively looking up the canonical element for the result of `Aliases.get_canonical_ignoring_name_mode` if it's not from the same compilation unit as the input.
I'll try to build a test case to see what happens in practice.

Reply:
I think we're fine actually, but the argument is not completely obvious.

The idea is that if `M.x` is known (in `M`) not to be canonical, then there shouldn't exist any occurrence of `M.x` outside the environment of `M` itself.
So if we get a reference to `M.x` without access to `m.cmx`, then it came (maybe transitively) from another imported unit that had access to `m.cmx`. I believe that this unit would never have kept an equation on `M.x` unless it saw it as canonical, but I will dig a bit more to get a concrete argument to support this belief.